### PR TITLE
Noise filter

### PR DESCRIPTION
**Pull request by luli-git**
fixed label selector and cluster builder

### CODE REVIEW

File: bin/experiment-utils.sh [hunk 1]
```diff
@@ -38,6 +38,7 @@ prepare_experiment_dir () {
 
     # Link in the vanilla vector file
     ln data/${map_name}/article_vectors.csv ${exp_dir}/vanilla_vectors.csv
+    ln data/${map_name}/article_labels.csv ${exp_dir}/chocolate_vectors.csv
```
Comment (shilad) [inline bin/experiment-utils.sh]: I don't think this is exactly right. I actually don't think the article_labels.csv needs to be copied at all. It can just be accessed from the map directory.

Comment (shilad) [inline bin/experiment-utils.sh]: I don't think this is exactly right. I actually don't think the article_labels.csv needs to be copied at all. It can just be accessed from the map directory.

File: bin/kmeans.sh [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            --articles_to_labels data/food/article_categories.csv \
+            --label_names data/food/category_names.csv
+
+    # Step 4: If you needed to generate augmented vectors,
+    # do so now from vanilla_vectors.csv in the experiment directory.
+
+    python -m cartograph.vector_augmenter \
+            --experiment ${exp_dir} \
+            --vectors ${exp_dir}/vanilla_vectors.csv \
+            --label_vectors ${exp_dir}/chocolate_vectors.csv \
```
Comment (shilad) [inline bin/kmeans.sh:54]: Get this directly from the map directory.

File: cartograph/cluster_builder.py [hunk 3]
```diff
@@ -7,12 +7,20 @@
...
+from scipy import spatial
 from sklearn.cluster import KMeans
 import pandas as pd
-from hdbscan import HDBSCAN
+import warnings
 import argparse
-
+import numpy as np
+# import numpy.linalg.norm
+with warnings.catch_warnings():
```
Comment (shilad) [inline cartograph/cluster_builder.py:18]: neat! i didn't know about this.

File: cartograph/cluster_builder.py [hunk 4]
```diff
@@ -21,34 +29,80 @@ def get_kmeans(article_vectors, k):
...
-
-
-def get_hdbscan(article_vectors, min_size):
+    centroids = pd.DataFrame(list(kmeans.cluster_centers_))
+    centroids.index.name = 'country'
+    df_with_vectors = pd.merge(df, article_vectors.drop(columns='country'), on='article_id')
+    df_with_centroids = pd.merge(df_with_vectors, centroids, on='country')
+    dist = []
+    for index, row in df_with_centroids.iterrows():
+        a = np.array(row[2:102])
```
Comment (shilad) [inline cartograph/cluster_builder.py]: it's better to not hard code the 100 here. calculate num_dims using the number of columns, or pass it in to the function.

File: cartograph/cluster_builder.py [hunk 5]
```diff
@@ -21,34 +29,80 @@ def get_kmeans(article_vectors, k):
...
+
+def normalize_vectors(dataframe):
+    mags = []
+    for index, row in dataframe.iterrows():
+        vector = row[1:100]
+        mag1 = mag(vector)
+        mags.append(mag1)
+    dataframe['magnitude'] = mags
+
+    try:
```
Comment (shilad) [inline cartograph/cluster_builder.py]: The divide by zero seems likely to happen :(. A better solution is leaving that vector zeros. You can do this by changing magnitudes of zero to 1.0

File: cartograph/cluster_builder.py [hunk 6]
```diff
@@ -21,34 +29,80 @@ def get_kmeans(article_vectors, k):
...
+    count = 0
+
+    if noise:
+        probabilities = hdbscan.probabilities_
+        df['country_likelihood'] = probabilities
+
+    for index, row in df.iterrows():
+        if row['country_likelihood'] == 0:
+            count = count + 1
+    print(count)
```
Comment (shilad) [inline cartograph/cluster_builder.py]: remove debugging.

File: cartograph/evaluation/modularity_evaluator.py [hunk 7]
```diff
@@ -23,13 +25,13 @@
...
 def preprocess(x_y_embeddings_csv):
-    df = pd.read_csv(x_y_embeddings_csv)
+    df = x_y_embeddings_csv
     feature_space = []
     indices_to_id = {}
 
     for index, row in df.iterrows():
         feature_space.append([row['x'], row['y']])
-        indices_to_id.update({index:int(row['article_id'])})
+        indices_to_id.update({index: int(row['article_id'])})
```
Comment (shilad) [inline cartograph/evaluation/modularity_evaluator.py]: `indices_to_id[index] = int(row['article_id'])`

File: cartograph/vector_augmenter.py [hunk 8]
```diff
@@ -65,21 +32,60 @@ def create_label_matrix(label_matrix):
...
+        cluster_df = pd.merge(article_vectors, label_vectors, on='article_id')
+    elif args.method == 'cluster':
+        cluster_df = pd.merge(cluster_vectors, article_vectors, on='article_id')
+    elif args.method == 'all':
+        cluster_df_with_cluster = pd.merge(cluster_vectors, article_vectors, on='article_id')
+        cluster_df = pd.merge(cluster_df_with_cluster, label_vectors, on='article_id')
+    else:
+        sys.stderr.write("Unkonwn clustering method: %s\n" + args.clustering)
+        sys.exit(1)
+    cluster_df.to_csv('%s/augmented_vectors.csv' % (args.experiment, ), index=False)
```
Comment (shilad) [inline cartograph/vector_augmenter.py]: I think this should take an output file rather than always writing to augmented_vectors.csv.

**Review by shilad — CHANGES_REQUESTED** [review]
Looking good! There is a ton of intricate work here! A few small changes, then I can merge.
